File: SmallStepSemantic/if.rb
class If < Struct.new(:condition, :consequence, :alternative)
  def to_s
    "if (#{condition}) {#{consequence}} else {#{alternative}}"
  end

  def inspect
    "#{self}"
  end

  def reducible?
    true
  end

  def reduce(environment)
    if condition.reducible?
      [If.new(condition.reduce(environment), consequence, alternative), environment]
    else
      case condition
      when Boolean.new(true)
        [consequence, environment]
      when Boolean.new(false)
        [alternative, environment]
      end
    end
  end
end

File: chapter02/BigStepSemantic/If.py
from chapter02.BigStepSemantic import Boolean


class If(object):
    def __init__(self, condition, consequence, alternative):
        self.condition = condition
        self.consequence = consequence
        self.alternative = alternative

    def evaluate(self, environment):
        if self.condition.evaluate(environment).value == Boolean(True).value:
            return self.consequence.evaluate(environment)
        elif self.condition.evaluate(environment).value == Boolean(False).value:
            return self.alternative.evaluate(environment)
